cpp/dolfinx/io/checkpointing.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2024 Abdullah Mujahid
+//
+// This file is part of DOLFINX (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#ifdef HAS_ADIOS2
+
+#include "checkpointing.h"
+#include "ADIOS2_utils.h"
+#include <adios2.h>
+#include <basix/finite-element.h>
+#include <dolfinx/mesh/Mesh.h>
+#include <mpi.h>
+
+using namespace dolfinx;
+using namespace dolfinx::io;
+
+/// @file checkpointing.h
+/// @brief ADIOS2 based checkpointing
+namespace
+{
+std::map<basix::element::lagrange_variant, std::string> lagrange_variants{
+    {basix::element::lagrange_variant::unset, "unset"},
+    {basix::element::lagrange_variant::equispaced, "equispaced"},
+    {basix::element::lagrange_variant::gll_warped, "gll_warped"},
+};
+
+template <std::floating_point T>
+void _write(ADIOS2Engine& adios2engine,
+            std::shared_ptr<dolfinx::mesh::Mesh<T>> mesh)
+{
+
+  auto io = adios2engine.io();
+  auto writer = adios2engine.engine();
+
+  const dolfinx::mesh::Geometry<T>& geometry = mesh->geometry();
+  auto topology = mesh->topology();
+
+  std::int16_t mesh_dim = geometry.dim();
+
+  auto imap = mesh->geometry().index_map();
+  std::uint64_t num_nodes_global = imap->size_global();
+  std::uint32_t num_nodes_local = imap->size_local();
+  std::uint64_t offset = imap->local_range()[0];
+
+  const std::shared_ptr<const dolfinx::common::IndexMap> topo_imap
+      = topology->index_map(mesh_dim);
+  std::uint64_t num_cells_global = topo_imap->size_global();
+  std::uint32_t num_cells_local = topo_imap->size_local();
+  std::uint64_t cell_offset = topo_imap->local_range()[0];
+
+  auto cmap = mesh->geometry().cmap();
```

Comment:
Do we have a clear idea of how this will extend to the mixed topology mesh case?

Reply:
@chrisrichardson can comment on the data structure of how mixed topology is implemented.

Similar to the example from meshtags, offsets and global size can be computed as

```
// Compute the global offset for owned (local) vertex indices
  std::uint64_t local_start = 0;
  {
    MPI_Exscan(&num_saved_tag_entities, &local_start, 1, MPI_UINT64_T, MPI_SUM,
               mesh->comm());
  }

  std::uint64_t num_tag_entities_global = 0;
  MPI_Allreduce(&num_saved_tag_entities, &num_tag_entities_global, 1,
                MPI_UINT64_T, MPI_SUM, mesh->comm());
```